### PR TITLE
Add new How-to article on automatic formatting of Flutter code

### PR DESCRIPTION
**Pull request by mit-mit**
Fixes https://github.com/flutter/flutter/issues/6418

### CODE REVIEW

**Comment by mit-mit** [comment]
Hi @sethladd, please review

**Comment by mit-mit** [comment]
We should remove https://github.com/flutter/flutter/wiki/Early-Access%3A-Auto-Formatting once this lands

File: formatting.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+sidebar: home_sidebar
+permalink: /formatting/
+---
+
+Flutter source code style guidelines, and support for automatic formatting.
+
+* TOC Placeholder
+{:toc}
+
+## Flutter formatting guidelines
```
Comment (sethladd) [inline formatting.md]: I think we can remove this from this article. It really belongs in the CONTRIBUTING.md or similar document. This isn't useful for most people, only really for contributors to the SDK itself.

File: formatting.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+While your code may follow any preferred style -- in our experience -- teams of
+developers may find it more productive to:
+
+1. have a single, shared style, and
+
+1. to enforce this style through automatic formatting.
+
+The alternative is often tiring formatting debates during code reviews, where
+time may be better spent on code behavior rather than code style. (*Note*:
+Automatically formatting code may in some hard cases lead to different results
```
Comment (sethladd) [inline formatting.md]: I think we can drop this parens. It's an interesting link but not in the critical path to enabling our developers to use the formatter.

File: formatting.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+that what you would personally expect. For background details, see our [blog
+post](
+http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/).)
+
+### Automatically formatting code in IntelliJ
+
+Automatic formatting of code is supported in IntelliJ if you have the
+`dart`-plugin (see [IntelliJ setup](/setup/#flutter-intellij-ide-plugins)).
```
Comment (sethladd) [inline formatting.md]: Not sure we need the `-` in `dart-plugin` ? Or, broaden the backticks? What is the formal name of the plugin?

File: formatting.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+To automatically format the code in the current source code window, right-click
```
Comment (sethladd) [inline formatting.md]: right click on what?

File: formatting.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+and select `Reformat with Dart style`. You can a keyboard shortcut to this in
```
Comment (sethladd) [inline formatting.md]: missing a word after "can" ?

File: formatting.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+Keymap section of IntelliJ Preferences.
+
+### Automatically formatting code in the CLI
```
Comment (sethladd) [inline formatting.md]: swap the acronym with the full words.

s/in the CLI/with command-line tools

File: formatting.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+You can also automatically format code in the CLI (command line interface) using
```
Comment (sethladd) [inline formatting.md]: s/in the CLI.../with the `flutter format` command.

File: formatting.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+the `flutter format` command:
+
+```
+Usage: flutter format <one or more paths>
+-h, --help    Print this usage information.
+```
+
+### Using 'trailing commas'
```
Comment (sethladd) [inline formatting.md]: I wonder if we should put this at top? This is the critical part.

File: formatting.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+To get good automatic formatting, we recommend you adopt the optional *trailing
+commas*. The rule for adding a trailing comma is simple: Always add a trailing
+comma at the end of a parameter list in functions, methods, and constructors
+where you care about keeping the formatting you crafted. This will help the
+automatic formatter to insert an appropriate amount of line breaks for
+Flutter-style code.
+
+Here is an example of automatically formatted code *with* trailing commas:
+
+![Automatically formatted code with trailing commas](/images/intellij/trailing-comma-with.png)
```
Comment (sethladd) [inline formatting.md]: When committing PNGs, please use a PNG optimizer like optipng (installable by homebrew)

File: formatting.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+```
+Usage: flutter format <one or more paths>
+-h, --help    Print this usage information.
+```
+
+### Using 'trailing commas'
+
+Flutter code often involves building fairly deep tree-shaped data structures,
+for example in a `build`-method. The automatic code formatting described above
```
Comment (sethladd) [inline formatting.md]: remove the -

File: formatting.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+is shared with other Dart code, and per default may lead to sub-optimal
+formatting of these deep structures.
```
Comment (sethladd) [inline formatting.md]: I think we can drop this last sentence. I think the main point for this section is:

* use trailing commas for better nesting
* here is a before and after. :)

File: formatting.md (hunk 1 above)
Comment (mit-mit) [inline formatting.md]: Done

File: formatting.md (hunk 3 above)
Comment (mit-mit) [inline formatting.md]: The formal name is `Dart` -- fixed the caps, but the tics are correct

File: formatting.md (hunk 5 above)
Comment (mit-mit) [inline formatting.md]: Fixed

File: formatting.md (hunk 6 above)
Comment (mit-mit) [inline formatting.md]: Done

File: formatting.md (hunk 7 above)
Comment (mit-mit) [inline formatting.md]: done

File: formatting.md (hunk 11 above)
Comment (mit-mit) [inline formatting.md]: Done

**Comment by sethladd** [comment]
Thanks for the contribution, but please wait for a LGTM from someone. Unless it's a truly trivial change, then we can do a TBR.

**Pull request merged**